renderapi/transform/leaf/thin_plate_spline.py

```diff
@@ -280,3 +286,72 @@ def dataString(self):
         b64_2 = encodeBase64(blk2)
 
         return '{} {} {}'.format(header, b64_1, b64_2)
+
+    def adaptive_mesh_estimate(
+            self,
+            src=None,
+            tol=1.0,
+            max_iter=50,
+            starting_grid=7,
+            nworst=10,
+            niter=0):
+        """method for creating a transform with fewer control points
+        that matches the original transfom within some tolerance.
+        Parameters
+        ----------
+        src : numpy.array
+            an Nx2 matrix of source points, None starts with grid
+        tol : float
+            in units of pixels, how close should the points match
+        max_iter: int
+            some limit on how many recursive attempts
+        starting_grid : int
+            estimate will start with an n x n grid
+        nworst : int
+            per iteration, the nworst matching srcPts will be added
+        niter : int
+            passed through the recursion for stopping criteria
+
+        Returns
+        -------
+        ThinPlateSplineTransform
+        """
+
+        if src is None:
+            mn = self.srcPts.min(axis=1)
+            mx = self.srcPts.max(axis=1)
+            xt, yt = np.meshgrid(
+                    np.linspace(mn[0], mx[0], starting_grid),
+                    np.linspace(mn[1], mx[1], starting_grid))
+            src = np.vstack((xt.flatten(), yt.flatten())).transpose()
+
+        dst = self.tform(src)
+        newtf = ThinPlateSplineTransform()
+        newtf.estimate(src, dst, computeAffine=(self.aMtx is not None))
+
+        if not hasattr(self, 'src'):
+            self.src = self.srcPts.transpose()
+        if not hasattr(self, 'dst'):
+            self.dst = self.tform(self.src)
+        newdst = newtf.tform(self.src)
+
+        delta = np.linalg.norm(newdst - self.dst, axis=1)
+        ind = np.argwhere(delta > tol).flatten()
+        sortind = np.argsort(delta[ind])
+
+        if niter == max_iter:
+            logger.warning(
```

Comment:
are we sure you want to not raise an exception here, seems the client of this should be in control of whether to accept the result and move on.  It could raise an Exception and return the newtf in the exception.

Reply:
also this should be tested.

Reply:
added the exception... not sure how to raise and return the newtf at the same time